replica/database.hh

```diff
@@ -857,7 +857,9 @@ public:
     db::replay_position set_low_replay_position_mark();
 
 private:
-    future<std::unordered_set<sstring>> take_snapshot(database& db, sstring jsondir);
```

Comment:
Why is copying sstring instances across shard fishy? They might result in an unnecessary copy, but I don't see anything fishy here. Also, this code doesn't seem to care for performance much, prioritizing ease of reading/writing, which is not helped by this foreign pointer.

Reply:
agree

Reply:
Currently the strings are allocated and freed on the calling coroutine frame
and we submit_to the "coordinator" shard to copy them across shard.

But the refactoring done here changes that, where the coordinator shard will submit_to all shards to perform `take_snapshot`, returning the `std::unordered_set<sstring>` in the future result.

We cannot free that on the coordinator shard anymore, so we need to wrap it in a foreign ptr so it will get freed on the shard that allocated it.

Looks like I should improve the explanation about the motivation for this patch...